enhancements/ingress/haproxy-router-compression.md

```diff
@@ -0,0 +1,421 @@
+
+**maxcompcpuusage**: sets the maximum CPU usage HAProxy Enterprise can reach before stopping the compression for
+new requests or decreasing the compression level of current requests. It works like maxcomprate but measures CPU
+usage instead of incoming data bandwidth. The value is expressed in percent of the CPU used by HAProxy Enterprise.
+In case of multiple processes (nbproc > 1), each process manages its individual usage. A value of 100 (the default)
+disables the limit. Setting a lower value will prevent the compression work from slowing the whole
+process down and from introducing high latencies.
+
+**tune.comp.maxlevel**: sets the maximum compression level. The compression level affects CPU usage during compression.
+Each session using compression initializes the compression algorithm with this value. The default is 1.
+
+### Open Questions
+### Test Plan
+- Unit tests will validate that HTTP compression directives appear in the `frontend`/`backend` sections of the HAProxy
+  template when it is properly configured in a route.
+- Unit tests will validate that the HTTP compression directive appears in the `defaults` section of the HAProxy template
+  when it is properly configured on the IngressController spec.
+- E2e tests will validate that the IngressController spec accepts and enables compression in the `defaults` section of
+  the generated haproxy.cfg
+
+Note that HAProxy Enterprise **disables** HTTP compression when:
+- The request does not advertise a supported compression algorithm in the "Accept-Encoding" header
+- The response message is not HTTP/1.1 or above
+- The HTTP status code is not one of 200, 201, 202 or 203
+- The response contains neither a "Content-Length" header, nor a "Transfer-Encoding" header whose last value is `chunked`
+- The response contains a "Content-Type" header whose first value starts with `multipart`
+- The response contains a "Cache-Control" header with a value of `no-transform`
+- The response contains a "Content-Encoding" header, indicating that the response is already compressed (see `compression offload`)
+- The response contains an invalid "ETag" header or multiple "Etag" headers
+
+### Graduation Criteria
+
+This feature will be immediately available in release 4.10, and will not have a dev or tech preview.
+
+#### Dev Preview -> Tech Preview
+n/a
+
+#### Tech Preview -> GA
+n/a
+
+#### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+An upgrade from a previous installation that had the compression directives exposed, will need to change to use the
+IngressController spec.  No previous installation would have any compression directives via route annotations.
+
+If a cluster administrator applied any of the compression configuration options
+in 4.10, then downgraded to 4.9, the compression configuration settings would no longer be in effect unless they
```

Comment:
Please be sure that there is an explicit test of this capability done by QE on rollback.

Reply:
Will add a unit test too.